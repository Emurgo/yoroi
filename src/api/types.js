// @flow

/**
 * API-related types
 */
import type {TransactionStatus} from '../types/HistoryTransaction'

export type RemoteAsset = {
  +amount: string,
  +assetId: string,
  +policyId: string,
  +name: string,
}
// this is equivalent to yoroi-frontend's `RemoteUnspentOutput`
export type RawUtxo = {|
  +amount: string,
  +receiver: string,
  +tx_hash: string,
  +tx_index: number,
  +utxo_id: string,
  +assets: $ReadOnlyArray<RemoteAsset>,
|}

export const CERTIFICATE_KIND = {
  STAKE_REGISTRATION: 'StakeRegistration',
  STAKE_DEREGISTRATION: 'StakeDeregistration',
  STAKE_DELEGATION: 'StakeDelegation',
  POOL_REGISTRATION: 'PoolRegistration',
  POOL_RETIREMENT: 'PoolRetirement',
  MOVE_INSTANTANEOUS_REWARDS: 'MoveInstantaneousRewardsCert',
}
export type CertificateKind = $Values<typeof CERTIFICATE_KIND>

// getAccountState

export type AccountStateRequest = {|
  addresses: Array<string>,
|}
export type RemoteAccountState = {|
  poolOperator: null, // not implemented yet
  remainingAmount: string, // current remaining awards
  rewards: string, // all the rewards every added
  withdrawals: string, // all the withdrawals that have ever happened
|}
export type AccountStateResponse = {|
  [key: string]: null | RemoteAccountState,
|}

// getPoolInfo

export type RemoteCertificate = {|
  kind: 'PoolRegistration' | 'PoolRetirement',
  certIndex: number,
  poolParams: Object, // don't think this is relevant
|}

export type RemotePoolMetaSuccess = {|
  info: ?{
    name?: string,
    ticker?: string,
    description?: string,
    homepage?: string,
    // other stuff from SMASH.
  },
  history: Array<{|
    epoch: number,
    slot: number,
    tx_ordinal: number,
    cert_ordinal: number,
    payload: RemoteCertificate,
  |}>,
|}

export type RemotePoolMetaFailure = {|
  error: Object,
|}

export type PoolInfoRequest = {|
  poolIds: Array<string>,
|}

export type PoolInfoResponse = {
  [key: string]: RemotePoolMetaSuccess | RemotePoolMetaFailure,
}

// getTxsBodiesForUTXOs

export type TxBodiesRequest = {|txsHashes: Array<string>|}
export type TxBodiesResponse = {[key: string]: string}

// reputation

export type ReputationObject = {
  node_flags?: number,
  // note: could be more metrics that are not handled
}
export type ReputationResponse = {[poolId: string]: ReputationObject}

// serverstatus

export type ServerStatusResponse = {|
  isServerOk: boolean,
  isMaintenance: boolean,
  serverTime: number, // in milliseconds
|}

// bestblock

export type BestblockResponse = {|
  height: number,
  epoch: ?number,
  slot: ?number,
  hash: ?string,
|}

// tx history

export type TxHistoryRequest = {|
  addresses: Array<string>,
  untilBlock: string,
  after?: {
    block: string,
    tx: string,
  },
|}

export type RemoteTransactionInputBase = {|
  +address: string,
  +amount: string,
  +assets: $ReadOnlyArray<RemoteAsset>,
|}

export type RemoteTransactionUtxoInput = {|
  +id: string, // concatenation of txHash || index
  +index: number,
  +txHash: string,
|}

// not considering acount txs for now
export type RemoteTransactionInput = {|
  ...RemoteTransactionInputBase,
  ...RemoteTransactionUtxoInput,
|}

export type RemoteTransactionOutput = {|
  +address: string,
  +amount: string,
  +assets: $ReadOnlyArray<RemoteAsset>,
|}

/**
 * only present if TX is in a block
 */
export type RemoteTxBlockMeta = {|
  +block_num: number,
  +block_hash: string,
  +tx_ordinal: number,
  +time: string, // timestamp with timezone
  +epoch: number,
  +slot: number,
|}

// See complete types in:
// https://github.com/Emurgo/yoroi-graphql-migration-backend#output-6
export type RemoteCertificateMeta =
  | {|
      kind: typeof CERTIFICATE_KIND.STAKE_REGISTRATION,
      rewardAddress: string, // hex
    |}
  | {|
      kind: typeof CERTIFICATE_KIND.STAKE_DEREGISTRATION,
      rewardAddress: string, // hex
    |}
  | {|
      kind: typeof CERTIFICATE_KIND.STAKE_DELEGATION,
      rewardAddress: string, // hex
      poolKeyHash: string, // hex
    |}
  | {|
      kind: typeof CERTIFICATE_KIND.POOL_REGISTRATION,
      poolParams: Object, // we don't care about this for now
    |}
  | {|
      kind: typeof CERTIFICATE_KIND.POOL_RETIREMENT,
      poolKeyHash: string, // hex
    |}
  | {|
      kind: typeof CERTIFICATE_KIND.MOVE_INSTANTANEOUS_REWARDS,
      rewards: {[addresses: string]: string},
      pot: 0 | 1,
    |}
export type RemoteTxInfo = {|
  +type: 'byron' | 'shelley',
  +fee?: string, // only in shelley txs
  +hash: string,
  +last_update: string, // timestamp with timezone
  +tx_state: TransactionStatus,
  +inputs: Array<RemoteTransactionInput>,
  +outputs: Array<RemoteTransactionOutput>,
  +withdrawals: Array<{|
    address: string, // hex
    amount: string,
  |}>,
  +certificates: Array<RemoteCertificateMeta>,
|}

export type RawTransaction = {|
  ...WithNullableFields<RemoteTxBlockMeta>,
  ...RemoteTxInfo,
|}
