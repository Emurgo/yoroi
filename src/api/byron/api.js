// @flow
import _ from 'lodash'
import {BigNumber} from 'bignumber.js'

import checkedFetch from '../fetch'
import {CONFIG} from '../../config/config'
import assert from '../../utils/assert'
import {checkAndFacadeTransactionAsync} from './facade'

import type {Transaction} from '../../types/HistoryTransaction'

import type {
  RawUtxo,
  TxBodiesRequest,
  TxBodiesResponse,
  ServerStatusResponse,
  BestblockResponse,
  TxHistoryRequest,
  AccountStateRequest,
  AccountStateResponse,
  PoolInfoRequest,
  PoolInfoResponse,
} from '../types'

const NETWORK_CONFIG = CONFIG.NETWORKS.HASKELL_SHELLEY.BACKEND

type Addresses = Array<string>

export const checkServerStatus = (): Promise<ServerStatusResponse> =>
  checkedFetch('status', null, NETWORK_CONFIG, 'GET')

export const getBestBlock = (): Promise<BestblockResponse> =>
  checkedFetch('v2/bestblock', null, NETWORK_CONFIG, 'GET')

export const fetchNewTxHistory = async (
  request: TxHistoryRequest,
): Promise<{isLast: boolean, transactions: Array<Transaction>}> => {
  assert.preconditionCheck(
    request.addresses.length <= NETWORK_CONFIG.TX_HISTORY_MAX_ADDRESSES,
    'fetchNewTxHistory: too many addresses',
  )
  const response = await checkedFetch('v2/txs/history', request, NETWORK_CONFIG)
  const transactions = await Promise.all(
    response.map(checkAndFacadeTransactionAsync),
  )
  return {
    transactions,
    isLast: response.length < NETWORK_CONFIG.TX_HISTORY_RESPONSE_LIMIT,
  }
}

export const filterUsedAddresses = async (
  addresses: Addresses,
): Promise<Addresses> => {
  assert.preconditionCheck(
    addresses.length <= NETWORK_CONFIG.FILTER_USED_MAX_ADDRESSES,
    'filterUsedAddresses: too many addresses',
  )
  // Take a copy in case underlying data mutates during await
  const copy = [...addresses]
  const used = await checkedFetch(
    'v2/addresses/filterUsed',
    {addresses: copy},
    NETWORK_CONFIG,
  )
  // We need to do this so that we keep original order of addresses
  return copy.filter((addr) => used.includes(addr))
}

export const fetchUTXOsForAddresses = (addresses: Addresses) => {
  assert.preconditionCheck(
    addresses.length <= NETWORK_CONFIG.FETCH_UTXOS_MAX_ADDRESSES,
    'fetchNewTxHistory: too many addresses',
  )
  return checkedFetch('txs/utxoForAddresses', {addresses}, NETWORK_CONFIG)
}

export const bulkFetchUTXOsForAddresses = async (
  addresses: Addresses,
): Promise<Array<RawUtxo>> => {
  const chunks = _.chunk(addresses, NETWORK_CONFIG.FETCH_UTXOS_MAX_ADDRESSES)

  const responses = await Promise.all(
    chunks.map((addrs) => fetchUTXOsForAddresses(addrs)),
  )
  return _.flatten(responses)
}

export const submitTransaction = (signedTx: string) => {
  return checkedFetch('txs/signed', {signedTx}, NETWORK_CONFIG)
}

export const fetchUTXOSumForAddresses = (
  addresses: Addresses,
): Promise<{sum: string}> => {
  assert.preconditionCheck(
    addresses.length <= NETWORK_CONFIG.FETCH_UTXOS_MAX_ADDRESSES,
    'fetchUTXOSumForAddresses: too many addresses',
  )
  return checkedFetch('txs/utxoSumForAddresses', {addresses}, NETWORK_CONFIG)
}

export const bulkFetchUTXOSumForAddresses = async (
  addresses: Addresses,
): Promise<{fundedAddresses: Array<string>, sum: BigNumber}> => {
  const chunks = _.chunk(addresses, NETWORK_CONFIG.FETCH_UTXOS_MAX_ADDRESSES)

  const responses = await Promise.all(
    chunks.map((addrs) => fetchUTXOSumForAddresses(addrs)),
  )
  const sum = responses.reduce(
    (x: BigNumber, y) => x.plus(new BigNumber(y.sum || 0)),
    new BigNumber(0),
  )

  const responseUTXOAddresses = await Promise.all(
    chunks.map((addrs) => fetchUTXOsForAddresses(addrs)),
  )

  const fundedAddresses = _.flatten(responseUTXOAddresses).map(
    (address: any) => address.receiver,
  )

  return {
    fundedAddresses: _.uniq(fundedAddresses),
    sum,
  }
}

export const getTxsBodiesForUTXOs = (
  request: TxBodiesRequest,
): Promise<TxBodiesResponse> => {
  return checkedFetch('txs/txBodies', request, NETWORK_CONFIG)
}

export const getAccountState = async (
  request: AccountStateRequest,
): Promise<AccountStateResponse> => {
  assert.preconditionCheck(
    request.addresses.length <= NETWORK_CONFIG.FETCH_UTXOS_MAX_ADDRESSES,
    'getAccountState: too many addresses',
  )
  return await checkedFetch('getAccountState', request, NETWORK_CONFIG)
}

export const bulkGetAccountState = async (
  addresses: Addresses,
): Promise<AccountStateResponse> => {
  const chunks = _.chunk(addresses, NETWORK_CONFIG.FETCH_UTXOS_MAX_ADDRESSES)

  const responses = await Promise.all(
    chunks.map((addrs) => getAccountState({addresses: addrs})),
  )
  return Object.assign({}, ...responses)
}

export const getPoolInfo = (
  request: PoolInfoRequest,
): Promise<PoolInfoResponse> => {
  return checkedFetch('getPoolInfo', request, NETWORK_CONFIG)
}
