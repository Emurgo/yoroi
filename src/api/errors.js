// @flow
import ExtendableError from 'es6-error'

// thrown when the request did go through but
// backend returned an unexpected result
export class ApiError extends ExtendableError {
  constructor(request: any) {
    super('ApiError')
    this.request = request
  }
}

// thrown when api failed to connect to the server
export class NetworkError extends ExtendableError {
  constructor() {
    super('NetworkError')
  }
}

// thrown by the backend after a rollback
export class ApiHistoryError extends ApiError {
  constructor(status: string) {
    super(`ApiHistoryError::${status}`)
  }
}
