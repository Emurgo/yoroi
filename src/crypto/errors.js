// @flow
import ExtendableError from 'es6-error'

export class CardanoError extends ExtendableError {}

export class WrongPassword extends ExtendableError {
  constructor() {
    super('WrongPassword')
  }
}

export class InsufficientFunds extends ExtendableError {
  constructor() {
    super('InsufficientFunds')
  }
}

export class NoOutputsError extends ExtendableError {
  constructor() {
    super('NoOutputsError')
  }
}

export const _rethrow = <T>(x: Promise<T>): Promise<T> =>
  x.catch((e) => {
    throw new CardanoError(e.message)
  })

export class InvalidState extends ExtendableError {}
