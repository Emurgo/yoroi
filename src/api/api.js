// @flow
import _ from 'lodash'
import {BigNumber} from 'bignumber.js'
import DeviceInfo from 'react-native-device-info'
import {Platform} from 'react-native'

import {Logger} from '../utils/logging'
import {CONFIG, CARDANO_CONFIG} from '../config'
import {NetworkError, ApiError} from './errors'
import assert from '../utils/assert'
import {checkAndFacadeTransactionAsync} from './facade'

import type {Moment} from 'moment'
import type {
  Transaction,
  RawUtxo,
  AccountStateResponse,
  PoolInfoRequest,
  PoolInfoResponse,
  TxBodiesRequest,
  TxBodiesResponse,
  ReputationResponse,
  ServerStatusResponse,
} from '../types/HistoryTransaction'

type Addresses = Array<string>

const _checkResponse = (response, requestPayload) => {
  if (response.status !== 200) {
    Logger.debug('Bad status code from server', response.status)
    Logger.debug('Request payload:', requestPayload)
    throw new ApiError(response)
  }
}

type RequestMethod = 'POST' | 'GET'

const _fetch = (
  path: string,
  payload: ?any,
  networkConfig: any,
  method?: RequestMethod = 'POST',
) => {
  const fullPath = `${networkConfig.API_ROOT}/${path}`
  const platform =
    Platform.OS === 'android' || Platform.OS === 'ios' ? Platform.OS : '-'
  const yoroiVersion = `${platform} / ${DeviceInfo.getVersion()}`
  Logger.info(`API call: ${fullPath}`)
  return (
    fetch(fullPath, {
      method,
      headers: {
        'Content-Type': 'application/json; charset=utf-8',
        'yoroi-version': yoroiVersion,
        'tangata-manu': 'yoroi',
      },
      body: payload != null ? JSON.stringify(payload) : undefined,
    })
      // Fetch throws only for network/dns/related errors, not http statuses
      .catch((e) => {
        Logger.info(`API call ${path} failed`, e)
        /* It really is TypeError according to
        https://developer.mozilla.org/en-US/docs/Web/API/Fetch_API/Using_Fetch
        */
        if (e instanceof TypeError) {
          throw new NetworkError()
        }
        throw e
      })
      .then(async (r) => {
        Logger.info(`API call ${path} finished`)

        _checkResponse(r, payload)
        const response = await r.json()
        // Logger.debug('Response:', response)
        return response
      })
  )
}

export const checkServerStatus = (
  networkConfig?: any = CONFIG.CARDANO,
): Promise<ServerStatusResponse> => _fetch('status', null, networkConfig, 'GET')

export const fetchNewTxHistory = async (
  dateFrom: Moment,
  addresses: Addresses,
  networkConfig?: any = CONFIG.CARDANO,
): Promise<{isLast: boolean, transactions: Array<Transaction>}> => {
  assert.preconditionCheck(
    addresses.length <= CONFIG.API.TX_HISTORY_MAX_ADDRESSES,
    'fetchNewTxHistory: too many addresses',
  )
  const response = await _fetch(
    'txs/history',
    {
      addresses,
      dateFrom: dateFrom.toISOString(),
    },
    networkConfig,
  )

  const transactions = await Promise.all(
    response.map(checkAndFacadeTransactionAsync),
  )
  return {
    transactions,
    isLast: response.length <= CONFIG.API.TX_HISTORY_RESPONSE_LIMIT,
  }
}

export const filterUsedAddresses = async (
  addresses: Addresses,
  networkConfig?: any = CONFIG.CARDANO,
): Promise<Addresses> => {
  assert.preconditionCheck(
    addresses.length <= CONFIG.API.FILTER_USED_MAX_ADDRESSES,
    'filterUsedAddresses: too many addresses',
  )
  // Take a copy in case underlying data mutates during await
  const copy = [...addresses]
  const used = await _fetch(
    'addresses/filterUsed',
    {addresses: copy},
    networkConfig,
  )
  // We need to do this so that we keep original order of addresses
  return copy.filter((addr) => used.includes(addr))
}

export const fetchUTXOsForAddresses = (
  addresses: Addresses,
  networkConfig?: any = CONFIG.CARDANO,
) => {
  assert.preconditionCheck(
    addresses.length <= CONFIG.API.FETCH_UTXOS_MAX_ADDRESSES,
    'fetchNewTxHistory: too many addresses',
  )
  return _fetch('txs/utxoForAddresses', {addresses}, networkConfig)
}

export const bulkFetchUTXOsForAddresses = async (
  addresses: Addresses,
  networkConfig?: any = CONFIG.CARDANO,
): Promise<Array<RawUtxo>> => {
  const chunks = _.chunk(addresses, CONFIG.API.FETCH_UTXOS_MAX_ADDRESSES)

  const responses = await Promise.all(
    chunks.map((addrs) => fetchUTXOsForAddresses(addrs, networkConfig)),
  )
  return _.flatten(responses)
}

export const submitTransaction = (
  signedTx: string,
  networkConfig?: any = CONFIG.CARDANO,
) => {
  return _fetch('txs/signed', {signedTx}, networkConfig)
}

export const fetchUTXOSumForAddresses = (
  addresses: Addresses,
  networkConfig?: any = CONFIG.CARDANO,
): Promise<{sum: string}> => {
  assert.preconditionCheck(
    addresses.length <= CONFIG.API.FETCH_UTXOS_MAX_ADDRESSES,
    'fetchUTXOSumForAddresses: too many addresses',
  )
  return _fetch('txs/utxoSumForAddresses', {addresses}, networkConfig)
}

export const bulkFetchUTXOSumForAddresses = async (
  addresses: Addresses,
  networkConfig?: any = CONFIG.CARDANO,
): Promise<{fundedAddresses: Array<string>, sum: BigNumber}> => {
  const chunks = _.chunk(addresses, CONFIG.API.FETCH_UTXOS_MAX_ADDRESSES)

  const responses = await Promise.all(
    chunks.map((addrs) => fetchUTXOSumForAddresses(addrs, networkConfig)),
  )
  const sum = responses.reduce(
    (x: BigNumber, y) => x.plus(new BigNumber(y.sum || 0)),
    new BigNumber(0),
  )

  const responseUTXOAddresses = await Promise.all(
    chunks.map((addrs) => fetchUTXOsForAddresses(addrs, networkConfig)),
  )

  const fundedAddresses = _.flatten(responseUTXOAddresses).map(
    (address: any) => address.receiver,
  )

  return {
    fundedAddresses: _.uniq(fundedAddresses),
    sum,
  }
}

export const getTxsBodiesForUTXOs = (
  request: TxBodiesRequest,
  networkConfig?: any = CONFIG.CARDANO,
): Promise<TxBodiesResponse> => {
  return _fetch('txs/txBodies', request, networkConfig)
}

export const fetchAccountState = async (
  addresses: Addresses,
  networkConfig?: any = CARDANO_CONFIG.SHELLEY,
): Promise<AccountStateResponse> => {
  assert.preconditionCheck(
    addresses.length <= CONFIG.API.FETCH_UTXOS_MAX_ADDRESSES,
    'fetchAccountState: too many addresses',
  )
  const response = await _fetch('v2/account/state', {addresses}, networkConfig)
  const mapped = {}
  for (const key of Object.keys(response)) {
    // Jormungandr returns '' when the address is valid but it hasn't appeared in the blockchain
    // edit: Jormungandr can now also return a description error when not in the blockchain
    if (response[key] === '' || response[key] === 'Account does not exist') {
      mapped[key] = {
        delegation: {pools: []},
        value: 0,
        counter: 0,
      }
    } else {
      mapped[key] = response[key]
    }
  }
  return mapped
}

export const getPoolInfo = (
  request: PoolInfoRequest,
  networkConfig?: any = CARDANO_CONFIG.SHELLEY,
): Promise<PoolInfoResponse> => {
  return _fetch('v2/pool/info', request, networkConfig)
}

export const getReputation = (
  networkConfig?: any = CARDANO_CONFIG.SHELLEY,
): Promise<ReputationResponse> => {
  return _fetch('v2/pool/reputation', null, networkConfig, 'GET')
}
