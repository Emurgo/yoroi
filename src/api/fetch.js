// @flow
import DeviceInfo from 'react-native-device-info'
import {Platform} from 'react-native'

import {Logger} from '../utils/logging'
import {NetworkError, ApiError, ApiHistoryError} from './errors'

import type {BackendConfig} from '../config/types'

type RequestMethod = 'POST' | 'GET'

const _checkResponse = async (rawResponse: Object, requestPayload: Object) => {
  let responseBody = {}
  try {
    responseBody = await rawResponse.json()
  } catch (_e) {
    throw new ApiError('unexpected server response')
  }
  const status = rawResponse.status
  if (status !== 200) {
    const resp = responseBody.error?.response
    if (
      resp === 'REFERENCE_TX_NOT_FOUND' ||
      resp === 'REFERENCE_BLOCK_MISMATCH' ||
      resp === 'REFERENCE_BEST_BLOCK_MISMATCH'
    ) {
      throw new ApiHistoryError(responseBody.error.response)
    }
    Logger.debug('Bad status code from server', status)
    Logger.debug('Request payload:', requestPayload)
    Logger.info('response', responseBody)
    throw new ApiError(responseBody.error?.response)
  }
  return responseBody
}

export default (
  path: string,
  payload: ?any,
  networkConfig: BackendConfig,
  method?: RequestMethod = 'POST',
  checkResponse?: (Object, Object) => Promise<Object> = _checkResponse,
) => {
  const fullPath = `${networkConfig.API_ROOT}/${path}`
  const platform =
    Platform.OS === 'android' || Platform.OS === 'ios' ? Platform.OS : '-'
  const yoroiVersion = `${platform} / ${DeviceInfo.getVersion()}`
  Logger.info(`API call: ${fullPath}`)
  return (
    fetch(fullPath, {
      method,
      headers: {
        'Content-Type': 'application/json; charset=utf-8',
        'yoroi-version': yoroiVersion,
        'tangata-manu': 'yoroi',
      },
      body: payload != null ? JSON.stringify(payload) : undefined,
    })
      // Fetch throws only for network/dns/related errors, not http statuses
      .catch((e) => {
        Logger.info(`API call ${path} failed`, e)
        /* It really is TypeError according to
        https://developer.mozilla.org/en-US/docs/Web/API/Fetch_API/Using_Fetch
        */
        if (e instanceof TypeError) {
          throw new NetworkError()
        }
        throw e
      })
      .then(async (r) => {
        Logger.info(`API call ${path} finished`)
        const response = await checkResponse(r, payload)
        // Logger.debug('Response:', response)
        return response
      })
  )
}
