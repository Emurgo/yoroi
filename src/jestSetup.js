// @flow
/* eslint-env jest */
/* This module sets up Jest */
import fetch from 'node-fetch'
import {Logger, LogLevel} from './utils/logging'

import nodeUtil from 'util'

jest.setMock('./config/networks', require('./config/__mocks__/networks'))
jest.mock('react-native-device-info', () => {
  return {
    getVersion: () => '1.5.1',
  }
})
jest.mock('react-native-randombytes', () => {
  const crypto = require('crypto')
  return crypto.randomBytes
})

jest.mock('@ledgerhq/react-native-hw-transport-ble', () => ({}))

// $FlowFixMe
global.TextEncoder = nodeUtil.TextEncoder
// $FlowFixMe
global.TextDecoder = nodeUtil.TextDecoder

export default {
  setup: () => {
    global.fetch = fetch
    Logger.setLogLevel(LogLevel.Warn)
  },
  debug: () => Logger.setLogLevel(LogLevel.Debug),
}
