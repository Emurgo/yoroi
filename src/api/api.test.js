// @flow
/* eslint-env jest */
import jestSetup from '../jestSetup'
import moment from 'moment'

import api from './'
import {ApiError} from './errors'

jestSetup.setup()
jest.setTimeout(30 * 1000)

describe('History API', () => {
  it('can fetch history', async () => {
    const addresses = [
      'Ae2tdPwUPEZKAx4zt8YLTGxrhX9L6R8QPWNeefZsPgwaigWab4mEw1ECUZ7',
    ]
    const ts = moment('1970-01-01')

    // We are async
    expect.assertions(1)
    const result = await api.fetchNewTxHistory(ts, addresses)

    expect(result.transactions[0]).toMatchSnapshot({
      bestBlockNum: expect.any(Number),
      lastUpdatedAt: expect.any(String), // these fields may change (e.g. after restarting a node)
      submittedAt: expect.any(String),
    })
  })

  it('throws ApiError on bad request', async () => {
    const addresses = []
    // mock moment
    const ts = {toISOString: () => 'not-a-date'}

    // We are async
    expect.assertions(1)

    await expect(api.fetchNewTxHistory(ts, addresses)).rejects.toThrow(ApiError)
  })

  it('filters used addresses', async () => {
    const addresses = [
      'Ae2tdPwUPEZKAx4zt8YLTGxrhX9L6R8QPWNeefZsPgwaigWab4mEw1ECUZ7',
      'Ae2tdPwUPEZ8wGxWm9VbZXFJcgLeKQJWKqREVEtHXYdqsqc4bLeGqjSwrtu',
      'Ae2tdPwUPEZ6T9qZxpao8ciAgg6ahjHRq2jV45ndZ4oPXAwrTYqN9NGUPh4',
      'Ae2tdPwUPEYzFrD5QSQ8NPHEJ8reecZkch5tT8wFAAYu18CJnkZ9XAm5ySE',
      'Ae2tdPwUPEZ3UJ2Y5zpbQeF8GEG1gtR1hoTdUoFJ5oHbbGkcinBrPJDhhhq',
      'Ae2tdPwUPEZ8ipN3wbA8heiZzuKzKRozK5SDzfuEjr8v38RRmFebaVN8ds6',
      'Ae2tdPwUPEZ1VJ4XrhmMMdyzKPmpEMcqajnt5HPCyAps5fF9aRFQCKYDCvj',
      'Ae2tdPwUPEZBr5uk5Rd1Ny2wVusVEqccTtsM5efBw6Sq1GdXiGzwAEDmjPP',
      'Ae2tdPwUPEYzX9Da4KhrarhhSaeFGB6bo8PByqmgj3TfJvr4jaouJ5Sns1N',
      'Ae2tdPwUPEZM6ok4jNYqLzU5Po2o68JdwNSspqai8axY4tJzYfQfj23M3vg',
      'Ae2tdPwUPEZ1WyLvNdf2wWPQgmACZ5ZgQBe6W8RYvNQsybYzQhAb5EFkwxQ',
    ]
    const used = [
      'Ae2tdPwUPEZKAx4zt8YLTGxrhX9L6R8QPWNeefZsPgwaigWab4mEw1ECUZ7',
      'Ae2tdPwUPEZ8wGxWm9VbZXFJcgLeKQJWKqREVEtHXYdqsqc4bLeGqjSwrtu',
      'Ae2tdPwUPEZ6T9qZxpao8ciAgg6ahjHRq2jV45ndZ4oPXAwrTYqN9NGUPh4',
      'Ae2tdPwUPEYzFrD5QSQ8NPHEJ8reecZkch5tT8wFAAYu18CJnkZ9XAm5ySE',
      'Ae2tdPwUPEZ3UJ2Y5zpbQeF8GEG1gtR1hoTdUoFJ5oHbbGkcinBrPJDhhhq',
      'Ae2tdPwUPEZ8ipN3wbA8heiZzuKzKRozK5SDzfuEjr8v38RRmFebaVN8ds6',
      'Ae2tdPwUPEZ1VJ4XrhmMMdyzKPmpEMcqajnt5HPCyAps5fF9aRFQCKYDCvj',
    ]

    expect.assertions(1)
    const result = await api.filterUsedAddresses(addresses)
    expect(result).toEqual(used)
  })

  it('keeps order in filterUsedAddresses', async () => {
    const addresses = [
      'Ae2tdPwUPEZKAx4zt8YLTGxrhX9L6R8QPWNeefZsPgwaigWab4mEw1ECUZ7',
      'Ae2tdPwUPEZ8wGxWm9VbZXFJcgLeKQJWKqREVEtHXYdqsqc4bLeGqjSwrtu',
      'Ae2tdPwUPEZ6T9qZxpao8ciAgg6ahjHRq2jV45ndZ4oPXAwrTYqN9NGUPh4',
      'Ae2tdPwUPEYzFrD5QSQ8NPHEJ8reecZkch5tT8wFAAYu18CJnkZ9XAm5ySE',
      'Ae2tdPwUPEZ3UJ2Y5zpbQeF8GEG1gtR1hoTdUoFJ5oHbbGkcinBrPJDhhhq',
      'Ae2tdPwUPEZ8ipN3wbA8heiZzuKzKRozK5SDzfuEjr8v38RRmFebaVN8ds6',
      'Ae2tdPwUPEZ1VJ4XrhmMMdyzKPmpEMcqajnt5HPCyAps5fF9aRFQCKYDCvj',
    ]

    expect.assertions(2)

    const result1 = await api.filterUsedAddresses(addresses)
    expect(result1).toEqual(addresses)

    addresses.reverse()

    const result2 = await api.filterUsedAddresses(addresses)
    expect(result2).toEqual(addresses)
  })
})
