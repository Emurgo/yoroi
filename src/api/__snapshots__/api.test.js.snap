// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`History API can fetch history 1`] = `
Object {
  "bestBlockNum": Any<Number>,
  "blockNum": 1889681,
  "id": "2401257195a3cd7d43e832b2d2e8215828ebbcdea472ef81f8de3314a535813e",
  "inputs": Array [
    Object {
      "address": "DdzFFzCqrhstJ7CLwbESFGX73gJw9inpvFnQjDn9QCpEXPUBdsw5NGHJnvwPB9Vkx6Jf6JPiWh9JQqwV9RDMfec3RSD968vWLwTNwRKE",
      "amount": "91951210",
    },
  ],
  "lastUpdatedAt": Any<String>,
  "outputs": Array [
    Object {
      "address": "Ae2tdPwUPEZKAx4zt8YLTGxrhX9L6R8QPWNeefZsPgwaigWab4mEw1ECUZ7",
      "amount": "5000000",
    },
    Object {
      "address": "DdzFFzCqrht3ZCWKPX3a1vHCjrhASfxdXMhdtus6eUcPhnB4o8pPdbggnuCM8dCyPhFnBtszUHkd8wtT4NVqEeypyTAiP2btobSNYC46",
      "amount": "86781326",
    },
  ],
  "status": "Successful",
  "submittedAt": Any<String>,
}
`;
